[workers] Made number of workers dynamic

ISauve opened the pull request:
### What does this PR do?

Makes the collector's runner have a dynamic number of check workers, based on the number of checks running -  if there are "too many" checks running extra workers get added (up to a maximum number). This can be overridden by the user to have a static (custom) number of check runners if desired.

### Motivation

The number of check workers should be dependant upon the number of check instances running. Previously, the number of check workers was static and user-specified, which means that if there are many checks required then manual configuration was required.

ISauve committed:
[workers] Made number of workers dynamic

Range-based number of workers, ranges configurable with check_workers.json

truthbk reviewed (changes requested):
Added some preliminary feedback regarding a few things we might want to do differently.

Also... tests! There's probably some testing you could write to test the actual scaling of the workers. Take a look at `pkg//collector/runner/runner_test.go` and maybe `pkg//collector/collector_test.go`.

truthbk reviewed (changes requested):
Added some notes

ISauve commented:
![](https://github.trello.services/images/mini-trello-icon.png) [[agent6] Adapt number of check workers to number of configured checks](https://trello.com/c/j5ucR2Wv/326-agent6-adapt-number-of-check-workers-to-number-of-configured-checks)

ISauve committed:
[workers] Tests: removed aggregator & reworked interface

Removed instantiating a nil aggregator for python tests, added
configuration via run-time flags, enabled lazy & busy wait for python
checks, refactored the way the time tests are run to clean up the code,
added configurability to the memory tests, added a README

truthbk commented on pkg/collector/runner/num_workers_test.go:
```diff
@@ -0,0 +1,227 @@
+
+	if runMultiple {
+		// To imitate a check running at an interval (UpdateNumWorkers doesn't run again)
+		for j := 0; j < numIntervals; j++ {
```
Why don't we just pass `numIntervals` as the parameter to `timeToComplete()` and drop `runMultiple`? I think it would make for the more natural interface. Otherwise we have this duplicated code, and nested code in this if, etc....

truthbk commented on pkg/collector/runner/num_workers_test.go:
```diff
@@ -0,0 +1,227 @@
+		t.Logf("%v\t%v\t%v\t%v\t%v\t", m.Alloc/1024, m.Sys/1024, m.HeapAlloc, m.HeapSys, m.HeapObjects)
+	}
+
+	for i := 1; i < 500; i++ {
```
I guess this upper bound could also be a parameter.

ISauve replied:
✅

truthbk commented on pkg/collector/runner/num_workers_test.go:
```diff
@@ -0,0 +1,227 @@
+const (
+	busyWait    CheckType = iota
+	lazyWait    CheckType = iota
+	pythonCheck CheckType = iota // BUG: testing the python checks crashes every so often
```
Also, what happened to the `BUG` in the comment? Do we no longer crash? If so, could we remove this comment?

ISauve replied:
Unfortunately, we do still crash occasionally... discussed with @masci and decided that it wasn't worth it to spend more time than we already have investigating the crash since I was able to run the tests enough times successfully to still be able to collect the data we wanted

truthbk commented on pkg/collector/runner/runner.go:
```diff
@@ -69,6 +90,46 @@ func NewRunner(numWorkers int) *Runner {
+	// Add workers if we don't have enough for this range
+	added := 0
+	for {
+		if numWorkers >= desiredNumWorkers {
```
Wondering if scaling down has an actual real-world use-case. Normally checks will be configured once and remain pretty static throughout the duration of the agent's life. That said, because of _AutoDiscovery_ there are cases where the number of configured checks could maybe go down(?). The overhead in memory for the additional go-routines is quite small, so I'm not sure how badly we'd want to implement the down-sizing.... It could maybe be implemented with a channel, or an atomic variable in the `Runner` class for the number of workers that should shutdown. We could discuss if you feel this makes sense.

ISauve replied:
That would be a nice feature, but I think for it to be really useful we would need there to be cases where there's at least 10 checks that get cancelled (otherwise the difference would be <=5 workers... not very significant), and I don't think that would happen often in an agents lifetime... not sure if it's worth implementing now? Definitely an option to keep in mind though!

truthbk commented on pkg/collector/runner/runner.go:
```diff
@@ -43,24 +52,36 @@ func init() {
-func NewRunner(numWorkers int) *Runner {
+func NewRunner() *Runner {
+	numWorkers := config.Datadog.GetInt("check_runners") // = 0 if no value is specified by the user
+	if numWorkers > maxNumWorkers {
```
we'd want to log something here if the configured value will not be enforced because it's higher than the `maxNumWorkers`

ISauve replied:
✅

truthbk commented on pkg/collector/runner/runner_test.py:
```diff
@@ -0,0 +1,11 @@
+    def check(self, instance):
+        # Busy wait for 100ms
+        current_time = time.time()
+        while (time.time() < current_time+0.1):
```
Could we configure the `100ms` busy wait using the `instance` configuration? That would give us added flexibility to mess around with testing parameters should we want to.

ISauve replied:
✅

truthbk reviewed (dismissed):
There's some really awesome stuff here, so cool you could do this and get such meaningful data! I added some feedback of things we could maybe cleanup, and maybe some ideas (we can discuss if you disagree, or if you'd like us to dive deeper, or if you think they're stupid - something perfectly viable :) )

Thanks!

ISauve commented on pkg/collector/runner/num_workers_test.go:
```diff
@@ -0,0 +1,227 @@
+
+/****************** Testing configuration ****************************/
+
+var testingEfficiency = false // Run this test (should be false normally)
```
Good idea - much cleaner. Done :+1:

ISauve commented on pkg/collector/runner/num_workers_test.go:
```diff
@@ -0,0 +1,227 @@
+
+	// Run the time tests
+	interval := false
+	for i := 0; i < 2; i++ {
```
The idea was that UpdateNumWorkers() only runs when a check gets added, not every time a check runs - so to simulate the same check running multiple times, the first time a check gets run it calls UpdateNumWorkers and then the next N times it doesn't call it... but I think doing it this way makes the code too unreadable, I'll change that

masci commented:
@ISauve @truthbk is this good to go?

ISauve commented on pkg/collector/runner/num_workers_test.go:
```diff
@@ -0,0 +1,227 @@
+	if !memoryTest {
+		return
+	}
+	// Run the memory test
```
I think the use case for this particular test would be as a standalone test, not running the entire test suite for the runner, which requires passing the name of the function you want to run as an argument to `go test`... so I think it would be a little awkward to have to pass 2 separate test names to pass... I can pull it into a separate function to clean up the code a little but I think it's still better left as a flag on this one test function. Thoughts?

ISauve commented on pkg/collector/runner/num_workers_test.go:
```diff
@@ -0,0 +1,227 @@
+		}
+		r.pending <- &c
+
+		if i%100 == 0 {
```
✅

ISauve commented on pkg/collector/runner/runner_test.py:
```diff
@@ -0,0 +1,11 @@
+
+class TestCheck(AgentCheck):
+    def check(self, instance):
+        # Busy wait for 100ms
```
✅

ISauve commented on pkg/collector/py/check.go:
```diff
@@ -89,6 +88,34 @@ func (c *PythonCheck) Run() error {
 }
 
+// RunSimple runs a Python check without sending data to the aggregator
+func (c *PythonCheck) RunSimple() error {
```
Agreed, it would definitely be ideal to mock the aggregator... originally, the aggregator was being initialized here https://github.com/DataDog/datadog-agent/blob/isabelle/workers/pkg/collector/runner/num_workers_test.go#L127
but @masci suggested that was one of the problems contributing to the python checks crashing so I removed it... not sure how else to handle this without changing the Run interface...

ISauve committed:
[workers] Limited configurability & modified interface

Made the number of checks configurable only in datadog.yaml, removed
argument from NewRunner

ISauve committed:
[workers] Added efficiency tests

masci reviewed (dismissed):
LGTM!
A couple of things I'd change (but can be done later): 1) keeping the global waitgroup `TestWg ` around only for the sake of testing doesn't seem quite right 2) we could move the perf test to a benchmark or a system test in the `/test/` folder at the root of the repo.

ISauve committed:
Merge branch 'master' into isabelle/workers
